src/main/java/com/spotify/sparkey/Util.java

```diff
@@ -266,41 +267,46 @@ static void nonThrowingClose(Closeable stream) {
     }
   }
 
-  static boolean renameFile(File srcFile, File destFile) throws FileNotFoundException {
+  static void renameFile(File srcFile, File destFile) throws IOException {
     if (!srcFile.exists()) {
       throw new FileNotFoundException(srcFile.getPath());
     }
     if (srcFile.equals(destFile)) {
-      return true;
+      return;
     }
     if (!destFile.exists()) {
-      return srcFile.renameTo(destFile);
+      rename(srcFile, destFile);
+      return;
     }
 
     File backupFile = new File(destFile.getParent(), destFile.getName() + "-backup" + UUID.randomUUID().toString());
     if (backupFile.exists()) {
       // Edge case, but let's be safe.
-      log.warn("Expected duplicate temporary backup file: " + backupFile.getPath());
-      return false;
-    }
-    if (!destFile.renameTo(backupFile)) {
-      return false;
+      throw new IOException("Expected duplicate temporary backup file: " + backupFile.getPath());
     }
+    rename(destFile, backupFile);
     if (destFile.exists()) {
       // This is strange, it should be renamed by now. But let's be safe.
-      log.warn("Unexpected file still existing: {}, should have been renamed to: {}", destFile.getPath(), backupFile.getPath());
-      return false;
+      throw new IOException("Unexpected file still existing: " + destFile.getPath() + ", should have been renamed to: " + backupFile.getPath());
     }
 
-    if (srcFile.renameTo(destFile)) {
-      boolean deleted = backupFile.delete();
-      return true;
+    try {
+      rename(srcFile, destFile);
+      if (!backupFile.delete()) {
+        log.warn("Could not delete backup file: " + backupFile.getPath());
+      }
+    } catch (IOException e) {
+      // roll back failed rename
+      try {
+        rename(backupFile, destFile);
```

Comment:
Move should be atomic, so this should not be necessary.

Reply:
This is to work around the case where we can't move A to B because B already exists.
Instead we do:
B -> C
A -> B
if success: delete C
if failure: C -> B

Reply:
You can overwrite files atomically though!